fix: #333 get rke2_node_name for only hosts in current cluster inventory group

michalg91 opened the pull request:
# Description
This fixes #333 

- The problem was in extracting hostvars by set_fact method which handled all hosts in inventory
- `register: node_names` registered ansible_fact dict so next steps in this block execution were comparing list to dict so were triggered when not needed
```
{'ansible_facts': {'node_names': ['node01', 'node02,  'node03']]}, 'failed': False, 'changed': False}
```


## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [ ] Small minor change not affecting the Ansible Role code (GitHub Actions Workflow, Documentation etc.)

## How Has This Been Tested?
Created clean nodes and set credentials for s3 restore, then run playbook which invoked role.

michalg91 committed:
fixes lablabs/ansible-role-rke2/#333 get rke2_node_name for only hosts in current cluster inventory group

michalg91 commented:
i encountered another problem (since this one encountered for me when i bumped role from 1.37.0 to 1.44.1)
i have automatic jenkins pipeline that restores cluster in daily manner, it brakes when joinining worker nodes now. 

https://github.com/lablabs/ansible-role-rke2/commit/e43df6b7c98e22ec1a796cf1796b4b6b6612338f this commit totally changed logic, so right after my "fix" it is not executed anymore since all hostnames have the same name in my backup scenario. 

@simonfelding do you remember why this logic is changed? i am able to fix it but i need to know what was the case here so my change will not produce any other failures.

simonfelding commented:
Hey, would love to help you out on this. Could you give us a bit more info on what exactly doesn't work? It's not entirely clear for me what it is your pipeline is doing and what the error is.

Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
From: michalg91 ***@***.***>
Sent: Tuesday, July 8, 2025 10:28:13 PM
To: lablabs/ansible-role-rke2 ***@***.***>
Cc: simonfelding ***@***.***>; Mention ***@***.***>
Subject: Re: [lablabs/ansible-role-rke2] fix: #333 get rke2_node_name for only hosts in current cluster inventory group (PR #334)

[https://avatars.githubusercontent.com/u/63045346?s=20&v=4]michalg91 left a comment (lablabs/ansible-role-rke2#334)<https://github.com/lablabs/ansible-role-rke2/pull/334#issuecomment-3050215027>

i encountered another problem (since this one encountered for me when i bumped role from 1.37.0 to 1.44.1)
i have automatic jenkins pipeline that restores cluster in daily manner, it brakes when joinining worker nodes now.

e43df6b<https://github.com/lablabs/ansible-role-rke2/commit/e43df6b7c98e22ec1a796cf1796b4b6b6612338f> this commit totally changed logic, so right after my "fix" it is not executed anymore since all hostnames have the same name in my backup scenario.

@simonfelding<https://github.com/simonfelding> do you remember why this logic is changed? i am able to fix it but i need to know what was the case here so my change will not produce any other failures.

—
Reply to this email directly, view it on GitHub<https://github.com/lablabs/ansible-role-rke2/pull/334#issuecomment-3050215027>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AKYOW7ZGPNSDBNPSN5KLQRT3HQSV3AVCNFSM6AAAAACBB6T6QOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTANJQGIYTKMBSG4>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

michalg91 commented:
Yeah, sure.
It works like that:
1. Login to control-plane node, trigger `rke2 etcd-snapshot save --name xxx`
2. Generates restore variables for role
3. Destroying cluster on **vmware** with `terraform destroy`
4. Recreates fresh nodes from same terraform definitions `terraform apply` (so here all node names are same like previous)
5. Runs `ansible-rke2-role` with provided restore settings

As i mentioned in issue, firstly it was broken in gathering `rke2_node_name` because it also tried to find `rke2_node_name` in other not related to `k8s_cluster` group hosts
https://github.com/lablabs/ansible-role-rke2/blob/602b420a9536724e066206b691f4b5292af39b24/tasks/first_server.yml#L169-L173
so i changed it to find in k8s cluster group, also removed register part because it causes to return dict instead of list of hosts. So diff always happened and it removed all my nodes secrets (also with fresh first master node).

After "my" change the step with removing "unwanted secrets" is skipped because no diff exists (which is true, all my nodes have same name as before destroy).

What i meant with changed logic. In 1.37.0 restore removed all nodes secrets in kube-system, but skipped master so playbook was working fine. Right now it tries to diff `kubectl get node` output with **ansible inventory nodes**. So in my case it is skipping deletion of old secrets and when node is joining to cluster it will not join because it has missmatched credentials `Node password rejected, duplicate hostname or contents of '/etc/rancher/node/password'`.

To fix this i think we should reintroduce previous version in just this step (deleting secrets) so all existing secrets instead currents playbook node will be deleted. The next step with deleting not existing nodes is fine for me with differential check.

michalg91 committed:
fix deleting secrets from old nodes

michalg91 commented:
https://github.com/lablabs/ansible-role-rke2/pull/334/commits/7518ec55c09f9672409fca1a8e94757921e577d8

this change fixed all my problems

michalg91 commented:
no @simonfelding i run some debug and output hostvars, the problem is what i stated and explained. When i am using inventory file with not only kubernetes hosts or running any play which invoked localhost before invoking role. It will not found `rke2_node_name` on host that is not part of cluster group. Just check it yourself :). Use `rke2_node_name` only in `k8s_cluster` group and debug this line by piping it to `ansible.builtin.debug`, with any any other host outside this group. 

Deleting secrets is not working with current version because it checks for difference with inventory file and kubectl output. In my case host names didn't change (there is no difference). So when playbook was joining reporovisioned nodes they missmatched rke2 credentials. So this step must delete old secrets - Just like before https://github.com/lablabs/ansible-role-rke2/commit/e43df6b7c98e22ec1a796cf1796b4b6b6612338f this commit. Also when you registered same variable name as fact name it dumps dict instead of list so after your commit difference always existed falsely.

Of course this can be also checked in debug.

Check it out.

put this inside debug.yaml to role
```yaml debug
 - name: Restore etcd
   when:
    - inventory_hostname == groups[rke2_servers_group_name].0
   block:
    - name: Get registered nodes
      ansible.builtin.shell:
        cmd: |
          set -o pipefail
          {{ rke2_data_path }}/bin/kubectl --kubeconfig /etc/rancher/rke2/rke2.yaml \
          get nodes --no-headers | awk '{print $1}'
      args:
        executable: /bin/bash
      changed_when: false
      register: registered_node_names

    - name: Get all node names
      ansible.builtin.set_fact:
        node_names: "{{ hostvars | dict2items | map(attribute='value.rke2_node_name') }}"
      run_once: true 
      register: node_names 

    - name: Remove old <node>.node-password.rke2 secrets
      ansible.builtin.debug: 
        msg: |
          {{ rke2_data_path }}/bin/kubectl --kubeconfig /etc/rancher/rke2/rke2.yaml \
          delete secret {{ item }}.node-password.rke2 -n kube-system 2>&1 || true
      # args:
      #   executable: /bin/bash
      with_items: "{{ registered_node_names.stdout_lines | difference(node_names) }}"
      changed_when: false

    - name: Remove old nodes
      ansible.builtin.debug:
        msg: |
          {{ rke2_data_path }}/bin/kubectl --kubeconfig /etc/rancher/rke2/rke2.yaml \
          delete node {{ item }} 2>&1 || true
      # args:
      #   executable: /bin/bash
      with_items: "{{ registered_node_names.stdout_lines | difference(node_names) }}"
      changed_when: false

    - name: variables
      ansible.builtin.debug:
        msg: |
          stdout
          {{ registered_node_names.stdout_lines }}

          node_names
          {{ node_names }}

          diffed
          {{ registered_node_names.stdout_lines | difference(node_names) }}

          diffed with node name
          {{ registered_node_names.stdout_lines | difference([rke2_node_name]) }}
```

and run:
```
- name: debug
  tags: debug
  gather_facts: yes
  become: yes
  hosts: k8s_cluster
  tasks:
    - ansible.builtin.include_role:
        name: lablabs.rke2
        tasks_from: debug
```

all in all recreating new cluster with etcd snapshot should delete all previously existed secrets so new joined node can join cluster.

simonfelding commented:
>all in all recreating new cluster with etcd snapshot should delete all previously existed secrets so new joined node can join cluster.

This is the part I'm worried about - is that always the case, or is that only in your case? You could also have a situation where you need to recover the etcd snapshot because etcd failed for some reason, but you didn't delete the nodes.

I'm not 100% sure why [rke2 uses the node-passwords like that, but maybe it's a security feature](https://docs.rke2.io/advanced#how-agent-node-registration-works)? I think we shouldn't just delete all the secrets just because etcd is being restored. 

https://github.com/lablabs/ansible-role-rke2/blob/602b420a9536724e066206b691f4b5292af39b24/tasks/first_server.yml#L171
It looks to me like the problem is actually just that line 171 assumes all hosts are rke2 nodes - which is why your localhost was included `"{{ hostvars | dict2items | map(attribute='value.rke2_node_name') }}"`.

Isn't the solution just to change it to `"{{ hostvars[groups['rke2_servers_group_name']] | dict2items | map(attribute='value.rke2_node_name') }}"`?

simonfelding commented:
Actually, the entire block Is misplaced.
https://github.com/lablabs/ansible-role-rke2/blob/602b420a9536724e066206b691f4b5292af39b24/tasks/first_server.yml#L155C1-L193C1

It doesn't do what it says it does (restore etcd), as this was already done earlier in the file. It actually deletes nodes that are no longer present in the ansible inventory.

I think this block belongs in the beginning of the `remaining_nodes.yml` file, to make it clear that it's a matter of removing nodes that aren't in the ansible inventory. It's not directly related to restoring etcd.
